Add a 'Post' control

kienstra opened the pull request:
* Right now, it's only available in the `/wp-admin` > Block Lab UI.
* The React control is not yet added, so it's not available in the block editor UI.

<img width="1150" alt="post-control-added" src="https://user-images.githubusercontent.com/4063887/54728064-180d6800-4b41-11e9-8057-66e6a1a21217.png">

Closes #19

kienstra committed:
Begin the 'Post' control, making it available

Right now, it's only available in the
/wp-admin > Block Lab UI.
The control is not yet added,
so it's not available in the block editor UI.

kienstra committed:
Add the Post control, very similar to the User one.

kienstra committed:
Demonstrate saving the ID, but displaying the title

This stores the post title in the displayValue state,
but there's a delay in loading it mounting the component.

kienstra committed:
Revert bfde7, store the displayValue in an attribute

It took too long to get the post title from the apiFetch
whenever the Post control mounted.
There probably needs to be a way to store the title (displayValue).
This doesn't seem ideal, but it removes the need to
call apiFetch to get the post title,
or the Taxonomy name for a future Taxonomy control.

kienstra commented:
**Storing Title In Attributes, Not Fetching It**

The [latest commit](https://github.com/getblocklab/block-lab/pull/257#issuecomment-482444369) stores the title [in the block's attributes](https://github.com/getblocklab/block-lab/pull/257/commits/529801ceb81b67ce05d356ad5cec7a810ad9d242#diff-445f4466528bc0064cf52d8c4c5f25beR38). It's not an ideal solution.

But it looks like fetching the title every time the components mounts might have too much of a delay.

This Post control needs to display the post's title in the `<input>`, but save the post's ID as the control value. The title is mutable, and the ID is not.

Before, I didn't see a clear way to store the post's title. So I used `apiFetch()` to get it when the component mounted:

![delayed-title](https://user-images.githubusercontent.com/4063887/56014397-d0e74280-5cba-11e9-8765-b9d3a9a19800.gif)

This is probably too long of a delay. Especially as this is in a local environment. It would probably be longer on a normal host.

This is still WIP, but it looks like storing the title [in the block's attributes](https://github.com/getblocklab/block-lab/pull/257/commits/529801ceb81b67ce05d356ad5cec7a810ad9d242#diff-445f4466528bc0064cf52d8c4c5f25beR38) gives a better user experience. It appears without the delay shown above.

kienstra committed:
Merge in develop, resolve conflict

The conflict was in:
test-class-block-post.php
It was trivial, simply from adding another assertion,
and modifying a comment.

kienstra committed:
Fix an issue where it didn't allow typing

Before, the onChange handler didn't account
for changing the <input> via the user typing into it.
Now, save that typing only to the displayValue.
Then, on choosing a post from the Popover,
it will set a new value with a post ID.

kienstra committed:
Improve JSDocs in the FetchInput component

Almost all of the functions were missing JSDocs.
So add them, to describe the component better.

kienstra committed:
Rename getDisplayValue to getDisplayValueFromAPI

It's more verbose,
but it's similar to getValueFromAPI,
as it's also getting a value from the API.

kienstra commented:
This is almost ready for review. The main use case is working, I just need to test it and fix any possible bugs.

lukecarbis commented:
@kienstra Looking great! What do you think about storing an _array_ instead of just the post title? It could include the post title, post type, and post ID.

lukecarbis commented on php/blocks/controls/class-post.php:
```diff
@@ -0,0 +1,80 @@
+				'default'  => '',
+				'sanitize' => 'sanitize_text_field',
+			)
+		);
```
It would be really cool to include a dropdown list of different post types here. It wouldn't have to be dynamic, as I think it's reasonable to assume the list of post types active on a given site isn't very long.

kienstra replied:
Good idea, @lukecarbis. Would this be a dropdown of all registered post types, like from [get_post_types()](https://codex.wordpress.org/Function_Reference/get_post_types) ?

lukecarbis replied:
Exactly, yes.

kienstra replied:
Sounds good, @lukecarbis. And it would be in the `/wp-admin` > Block Lab UI, not the block editor, right?

![dropdown-post-types](https://user-images.githubusercontent.com/4063887/56170814-94f8f980-5fa8-11e9-8db6-693d47a9e6b9.jpg)

lukecarbis replied:
Yes – exactly. Then in the block editor the posts would be filtered by the post type you selected here.

kienstra replied:
Good idea. This is implemented now.

kienstra commented:
**Storing Array**

Hi @lukecarbis,
>@kienstra Looking great! What do you think about storing an array instead of just the post title? It could include the post title, post type, and post ID.

Great idea. I'll work on implementing that.

kienstra committed:
Store the Post control value as an Object, not a string

Before, this was a Number, and there was a separate
string for the displayValue (title).
Thanks to Luke's suggestion, simply save this as an Object,
and store both the post ID and title in that.

kienstra committed:
Change the validate() method, update unit tests

Change this method to keep up with the changes
in how the attributes are stored as an array.

kienstra committed:
In the /wp-admin > Block Lab UI, add a post_type <select>

For the Post control, this will allow choosing
a public post type.

kienstra committed:
Use the rest_base for the post_type, to make fetches possible

Before, this used the slug of the post type.
But that sometimes doesn't help in doing
API requests for it.

kienstra committed:
Pass the newly-create post_type field to the Post control

It looks to work now,
and searches for the specific post type.

kienstra committed:
Improve DocBlocks and parameter names

Use apiResponse as the parameter name,
as it's more descriptive than value.

kienstra commented:
**Maybe This Shouldn't Include Attachments**

This now allows selecting all public post types. That includes media (attachments). 

But maybe this shouldn't include media, as there's an Image control for that. 

<img width="968" alt="post-control" src="https://user-images.githubusercontent.com/4063887/56184590-33548180-5fe0-11e9-83b2-b25e0859e061.png">

kienstra committed:
Remove extra spaces

attributes.js shouldn't be part of this diff at all,
and remove an extra space between ' and )

kienstra commented:
**Request For Review**

Hi @lukecarbis,
Could you please review this? It has the `<select>` element to query by post type. For example, if you select 'Pages', it will only query for pages:

![page-post-type](https://user-images.githubusercontent.com/4063887/56185121-36e90800-5fe2-11e9-940d-653312d380b0.gif)

lukecarbis commented on php/blocks/controls/class-control-abstract.php:
```diff
@@ -388,4 +418,27 @@ function( $option ) use ( &$options ) {
+				||
+				$post_type === $value
+			) {
+				return $value;
```
This looks like it will sort-of fail to sanitise (or at least fail to return a value) if the `$value` passed in isn't a registered post type, _or_ if there are no registered post types. This could be refactored so that we return `null` or `false` or `''` or something if there's a bad result.
e.g.
```
	public function sanitize_post_type( $value ) {
		$value = '';

		$public_post_types = get_post_types( array( 'public' => true ) );
		foreach ( $public_post_types as $post_type ) {
			$post_type_object = get_post_type_object( $post_type );

			if ( $post_type_object && ( $value === $post_type_object->rest_base ) ) {
				$value = $post_type_object->rest_base;
				break;
			}

			if ( $value === $post_type ) {
				$value = $post_type;
				break;
			}
		}

		return $value;
	}
```

But maybe there's a better way?

kienstra replied:
Hi @lukecarbis,
Good point that it doesn't return anything here. I wasn't quite sure what was best.

But it looks like having the validation method return `null` (or not return anything) is necessary to fall back to [the default value](https://github.com/getblocklab/block-lab/blob/ccabbed990b5541e5ac3ff59b1352c6d8bbe3dad/php/blocks/controls/class-control-setting.php#L114). So if the validation fails somehow, this will query for the post type of `post` as expected.

I could see how it's strange that the function sometimes doesn't return anything. But it [looks like](https://www.php.net/manual/en/functions.returning-values.php) not returning anything means that it returns `null`:
>If the return is omitted the value NULL will be returned.

Not a strongly-held opinion, though.

kienstra replied:
On second thought, I think it is better to have a return statement at the end of the function in case it hasn't returned. I'll make a commit with that.

kienstra replied:
https://github.com/getblocklab/block-lab/pull/257/commits/95cc6d2705959fee5f5b72b7646068657ffb146c#diff-dc4ef2232132f132dbd59b26a1f8db80R454 returns `null` at the end of the function. I also refactored it so there's only one method that finds all of the eligible post types.

lukecarbis reviewed (changes requested):
Very nicely done! Love the solutioning and architecture behind this.

I noticed that the post type isn't saved in the JSON, only the post title and ID. I can't think of a good reason to store the post type as well, since it's pretty accessible via the post ID, but we had talked about including it. Happy to leave it out if that's what you think is best.

```
<!-- wp:block-lab/post-block
{
    "text-only": "Test",
    "choose-a-page": {
        "id": 2,
        "title": "Sample Page"
    },
    "more-text": "More text",
    "inspector-post": {
        "id": 1,
        "title": "Hello world!"
    },
    "user": "lukecarbis"
}
/-->
```

If it's an easy fix, notice that the user input only saves the username, not the ID? I think it would be better if it matched the Post control. Would you prefer to fix that here or in a separate PR?

Another inconsistency with the User control is that if you type in _invalid_ username, and then blur the field, the field displays an error. The Post control clears the text (in addition to a red border to indicate that there was a problem). I'm not sure if this was intentional, but I prefer the latter (Post style) way of handling invalid text. Could we make this consistent?

Lastly, I agree that we should filter out the Attachment post type. This shouldn't be an option.

RobStino commented:
Great work fellas!
Testing now. Spotted a couple of things and possibly a bug:
1. If a page/post has no title it can't be selected, even if it's published. Edge-case issue.
2. The filtered select experience feels a little wrong when you want to re-select. Not a bug, bust something to think about in general
3. I was missing a page (original Sample page) in my select options. This might be a bug.

These 3 points are covered in this screencap.

https://d.pr/v/wTI3HT

kienstra committed:
Change the field to 'post_type_rest_slug'

It's actually the REST slug, not the post type.
A CPT can register a rest_base.
For example, if the post_type is testimonial,
it can have a rest_base of testimonials.
So this has to account for that.

kienstra commented:
Hi @lukecarbis,
>I noticed that the post type isn't saved in the JSON, only the post title and ID. I can't think of a good reason to store the post type as well, since it's pretty accessible via the post ID, but we had talked about including it. Happy to leave it out if that's what you think is best.

Yeah, it probably wouldn't hurt to store the `post_type` separately. But `block_value( 'foo-field' )` will return a `WP_Post` object, so they could access it with the `->post_type` property, like you mentioned.

kienstra committed:
Add return null at the end of sanitize_post_type_rest_slug()

As Luke mentioned, it should return something.
Also, refactor this to use a common method.

kienstra commented:
**Sample Page**

Hi @RobStino,
Thanks a lot for testing this and for your thoughts.

It looks like the "Sample Page" is published in the [screencast](https://d.pr/v/wTI3HT). So I'm not sure why it doesn't appear.

It looks to appear in my local:
![sample-page-displays](https://user-images.githubusercontent.com/4063887/56377834-46688b00-6246-11e9-905d-1020fb4100af.gif)

kienstra commented:
Hi @lukecarbis,
Thanks a lot for reviewing this.

>If it's an easy fix, notice that the user input only saves the username, not the ID? I think it would be better if it matched the Post control. Would you prefer to fix that here or in a separate PR?

Sure, would you mind if I looked at this in another PR? It would have to migrate the values from the previous `string` to an `array()`. Maybe that won't be too complex, though.

>Another inconsistency with the User control is that if you type in invalid username, and then blur the field, the field displays an error. The Post control clears the text (in addition to a red border to indicate that there was a problem). I'm not sure if this was intentional, but I prefer the latter (Post style) way of handling invalid text. Could we make this consistent?

If it's alright, I'll work on this tomorrow.

>Lastly, I agree that we should filter out the Attachment post type. This shouldn't be an option.

Sounds good, https://github.com/getblocklab/block-lab/pull/257/commits/95cc6d2705959fee5f5b72b7646068657ffb146c#diff-dc4ef2232132f132dbd59b26a1f8db80R292 removes the Attachment post type.

kienstra commented:
Hi @RobStino,
>1. If a page/post has no title it can't be selected, even if it's published. Edge-case issue.

Good catch. I'll work on this tomorrow if that's alright.

>2. The filtered select experience feels a little wrong when you want to re-select. Not a bug, bust something to think about in general

Yeah, it is a little strange. It probably needs more thought.

kienstra commented:
**If There's No Post Title**

Hi @RobStino,
Thanks for bringing up that you can't select a post if it has no title.

How about not even fetching posts without titles, so the user can't even select them from the Popover?

This control outputs the title for `block_field( 'foo-post' )`. So maybe it's not meaningful to be able to select a post without a title. Though `block_value( 'foo-post' )` exposes the `WP_Post`.

kienstra committed:
On tabbing away from user field with no exact match, clear the <input>

As Luke mentioned, this is the behavior of the Post control.
So make this consistent in the User control.

kienstra committed:
If the <input> value is invalid, don't add anything

But still allow selecting a value
if the <input> is valid.

kienstra committed:
Remove the TAB handler, as onBlur seems to handle that

kienstra committed:
Don't display results at all if they have no title

It doesn't seem useful to be able to select something
without a title,
as this will usually be what echoes on block_field().

kienstra commented:
Hi @lukecarbis,
>Another inconsistency with the User control is that if you type in invalid username, and then blur the field, the field displays an error. The Post control clears the text (in addition to a red border to indicate that there was a problem). I'm not sure if this was intentional, but I prefer the latter (Post style) way of handling invalid text. Could we make this consistent?

Good catch, [these commits](https://github.com/getblocklab/block-lab/compare/95cc6d2705959fee5f5b72b7646068657ffb146c...f18216388a036db41ea24cbcb72be12ef0ab01bb) clear the User control when the `<input>` is invalid:

![user-control](https://user-images.githubusercontent.com/4063887/56409197-eb20b200-62b2-11e9-8d7b-f325e39516e5.gif)

kienstra commented:
https://github.com/getblocklab/block-lab/pull/257/commits/aef4922bcdf7d59c066595473e14a05de8f9a5b1 prevents any result from displaying if there's no title. This is open for discussion, please see https://github.com/getblocklab/block-lab/pull/257#issuecomment-484757502

kienstra commented:
**Request For Another Round Of Review**

Hi @lukecarbis,
Thanks for waiting. Could you please review this again when you can?

It address all of the points except:

>If it's an easy fix, notice that the user input only saves the username, not the ID? I think it would be better if it matched the Post control. Would you prefer to fix that here or in a separate PR?

This requires a little thought about backwards-compatibility, so it'd be great if I could address this in another PR.

>The filtered select experience feels a little wrong when you want to re-select. Not a bug, bust something to think about in general

Maybe we could also consider this separately.

lukecarbis commented:
See #288.

lukecarbis commented on php/blocks/controls/class-control-abstract.php:
```diff
@@ -281,6 +275,31 @@ public function render_settings_post_type_rest_slug( $setting, $name, $id ) {
+	 *     @type string $name The name of the post type.n
+	 * }
+	 */
+	public function get_post_type_rest_slugs() {
```
Great idea separating this out into its own method.

kienstra replied:
Thanks, Luke!

lukecarbis commented on php/blocks/controls/class-control-abstract.php:
```diff
@@ -429,16 +448,9 @@ function( $option ) use ( &$options ) {
+		if ( array_key_exists( $value, $this->get_post_type_rest_slugs() ) ) {
+			return $value;
 		}
+		return null;
```
👌

lukecarbis committed:
Separate logic for skipping attachment posts when retrieving post type slugs

lukecarbis reviewed (dismissed):
@kienstra Code looks good. For anyone else testing, I had to resave the block I created for testing earlier in this PR, making sure to reselect the correct Post Type.

I want to make one change before we merge this.

lukecarbis commented:
Actually, never mind! Merge away.

I was hoping to add an "Any Post Type" as the default for the Post Type option, but only just now realised that there's no REST API endpoint for querying all post types at once.

kienstra committed:
Change title in attributes to name

This will probably share a common file with the
Taxonomy control.
And the taxonomy has a name, not a title.

kienstra commented:
**Made one last commit**

Hi @lukecarbis, sorry to do this last-minute. But I changed `'title'` to `'name'` in the Post control's attributes. 

I've been working on the Taxonomy control, and it will probably share most of the logic with this Post controls.

And I thought `'name'` would be more appropriate than `'title'`, as taxonomies only have a `'name'`.

lukecarbis reviewed (approved):
👍
